File: GestionAbsence-1/src/main/java/com/enicarthage/services/CoursService.java
package com.enicarthage.services;

import com.enicarthage.entities.Cours;
import com.enicarthage.entities.Groupe;
import com.enicarthage.entities.Professeur;
import com.enicarthage.entities.SessionCours;
import com.enicarthage.repositories.CoursRepository;
import com.enicarthage.repositories.GroupeRepository;
import com.enicarthage.repositories.ProfesseurRepository;
import com.enicarthage.repositories.SessionCoursRepository;

import jakarta.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class CoursService {

    @Autowired
    private CoursRepository coursRepository;
    @Autowired
    private GroupeRepository groupeRepository;
    @Autowired
    private SessionCoursRepository sessioncoursRepository;
    @Autowired
    private ProfesseurRepository professeurRepository;

    public List<Cours> getAllCours() {
        return (List<Cours>) coursRepository.findAll();
    }

    public Cours getCoursByNom(String nomCours) {
        return coursRepository.findById(nomCours).orElse(null);
    }
    public Cours createCours(Cours cours) {
        if (cours.getGroup() == null || cours.getGroup().getNom() == null) {
            throw new IllegalArgumentException("Cours must have a valid group associated with it");
        }
        
        Groupe groupe = groupeRepository.findByNom(cours.getGroup().getNom());
        if (groupe == null) {
            throw new IllegalArgumentException("Group with name " + cours.getGroup().getNom() + " not found");
        }
        cours.setGroup(groupe);
        
        return coursRepository.save(cours);
    }


    public Cours updateCours(String nomCours, Cours cours) {
        if (coursRepository.existsById(nomCours)) {
            cours.setNomCours(nomCours);
            return coursRepository.save(cours);
        } else {
            return null; // or throw exception
        }
    }

    public void deleteCours(String nomCours) {
        coursRepository.deleteById(nomCours);
    }

    @Transactional
    public SessionCours addSessionCours(String nomCours, SessionCours sessionCours) {
        Cours cours = coursRepository.findById(nomCours)
                .orElseThrow(() -> new RuntimeException("Cours not found with name: " + nomCours));
        sessionCours.setCours(cours);
        return sessioncoursRepository.save(sessionCours);
    }
    
    public Cours getCoursBySessionId(Long sessionId) {
        return coursRepository.findBySessionsIdSession(sessionId);
    }
    
    public List<Cours> getCoursByProfesseurId(Long professeurId) {
        return coursRepository.findByIdProfesseur(professeurId);
    }
    
    public Professeur getProfesseurByNomCours(String nomCours) {
        Cours cours = coursRepository.findByNomCours(nomCours);
        if (cours != null) {
            return professeurRepository.findById(cours.getIdProfesseur()).orElse(null);
        } else {
            return null;
        }
    }
    public List<SessionCours> getSessionCoursByNomCours(String nomCours) {
        return coursRepository.findByNomCours(nomCours).getSessions();
    }
    
}

File: Front/src/app/services/cours.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Cours } from '../models/Cours';
import { SessionCours } from '../models/SessionCours';
import { Professeur } from '../models/Professeur';

@Injectable({
  providedIn: 'root'
})
export class CoursService {

  private apiUrl = 'http://localhost:8081/cours';

  constructor(private http: HttpClient) { }

  getAllCours(): Observable<Cours[]> {
    return this.http.get<Cours[]>(`${this.apiUrl}`);
  }

  getCoursByNom(nomCours: string): Observable<Cours> {
    return this.http.get<Cours>(`${this.apiUrl}/${nomCours}`);
  }

  createCours(cours: Cours): Observable<Cours> {
    return this.http.post<Cours>(`${this.apiUrl}`, cours);
  }

  updateCours(nomCours: string, cours: Cours): Observable<Cours> {
    return this.http.put<Cours>(`${this.apiUrl}/${nomCours}`, cours);
  }

  deleteCours(nomCours: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${nomCours}`);
  }

  addSessionCours(nomCours: string, sessionCours: SessionCours): Observable<SessionCours> {
    return this.http.post<SessionCours>(`${this.apiUrl}/${nomCours}/sessions`, sessionCours);
  }

  getCoursBySessionId(sessionId: number): Observable<Cours> {
    return this.http.get<Cours>(`${this.apiUrl}/bySessionId/${sessionId}`);
  }

  getCoursByProfesseurId(professeurId: number): Observable<Cours[]> {
    return this.http.get<Cours[]>(`${this.apiUrl}/byProfesseurId/${professeurId}`);
  }

  getProfesseurByNomCours(nomCours: string): Observable<Professeur> {
    return this.http.get<Professeur>(`${this.apiUrl}/professeur/${nomCours}`);
  }

  getSessionCoursByNomCours(nomCours: string): Observable<SessionCours[]> {
    return this.http.get<SessionCours[]>(`${this.apiUrl}/${nomCours}/sessions`);
  }

}
